infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/server/KubernetesServerExposer.java

```diff
@@ -175,57 +172,89 @@ public void expose(Map<String, ? extends ServerConfig> servers) throws Infrastru
     Map<String, ServerConfig> internalServers = new HashMap<>();
     Map<String, ServerConfig> externalServers = new HashMap<>();
     Map<String, ServerConfig> secureServers = new HashMap<>();
+    Map<String, ServicePort> servicePorts = new HashMap<>();
+    Map<String, ServicePort> securePorts = new HashMap<>();
 
     servers.forEach(
         (key, value) -> {
+          ServicePort sp = getServicePort(value);
+          exposeInContainerIfNeeded(sp);
           if (value.isInternal()) {
             // Server is internal. It doesn't make sense to make an it secure since
             // it is available only within workspace servers
             internalServers.put(key, value);
+            servicePorts.put(value.getPort(), sp);
           } else {
             // Server is external. Check if it should be secure or not
             if (value.isSecure()) {
               secureServers.put(key, value);
+              securePorts.put(value.getPort(), sp);
             } else {
               externalServers.put(key, value);
+              servicePorts.put(value.getPort(), sp);
             }
           }
         });
 
-    Collection<ServicePort> servicePorts = exposePorts(servers.values());
-    Service service =
-        new ServerServiceBuilder()
-            .withName(generate(SERVER_PREFIX, SERVER_UNIQUE_PART_SIZE) + '-' + machineName)
-            .withMachineName(machineName)
-            .withSelectorEntry(CHE_ORIGINAL_NAME_LABEL, pod.getMetadata().getName())
-            .withPorts(new ArrayList<>(servicePorts))
-            .withServers(internalServers)
-            .build();
+    Map<String, ServerConfig> allNonSecureServers = new HashMap<>(internalServers);
+    allNonSecureServers.putAll(externalServers);
+    // only explicitly expose the service for the non-secure servers
+    if (!servicePorts.isEmpty()) {
```

Comment:
can we extract this block to separate method? It's very hard to follow such long method doing so many things